design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+## Approach
+
+There is a wide spectrum of complexity between minimally useful row lookups and arbitrarily complex SQL queries over materialized state. Adding support for even the most trivial queries will still add significant value for KSQL users, so we should consider interactive queries an ongoing, incremental effort spanning numerous KSQL releases.
+
+Initial limitations around the kinds of interactive queries KSQL can support will also be augmented by the flexibility that users still have when defining materialized views with continuous queries. For example, even if aggregates are not initially supported for interactive queries, they can still be used in a materialized view’s definition. The same is true for joins.
+
+Here we roughly subdivide the spectrum of interactive query complexity into five progressive tiers, each tier building on the one before it. Each of these tiers can be released separately, although it may make sense to combine some of them.
+
+*Please note that some of the example queries below use mock functions for illustrative purposes.*
+
+### Tier 0 - Simple key-based lookups
+
+Support for key-based lookups is the bare minimum starting point for interactive queries that will still add meaningful utility. This essentially amounts to a cosmetic SQL layer above RocksDB’s key-value interface. Key-based lookups will allow users to look up a single row, given a row key:
+
+* WHERE clause must be strictly of the form: `WHERE ROWKEY = <rowkey>`
+* Each query can therefore only target a single shard.
+* Arbitrary expressions should be allowed.
+* Examples:
+  * `SELECT x, y, z FROM table WHERE ROWKEY = ‘rowkey’`
+  * `SELECT x + y FROM table WHERE ROWKEY = ‘rowkey’`
+
+### Tier 1 - Key-based range queries
+
+Key-based range queries are a minimal incremental step forward from single-shard queries towards multi-shard queries. Range queries may target multiple shards, but restricting them to be key based will allow us to leverage RocksDB’s native range scan support instead of doing extraneous row scanning:
+
+* Queries with a predicate of the form WHERE ROWKEY > lower_bound AND ROWKEY < upper_bound should be allowed.
+* While not technically a range query, queries with a WHERE ROWKEY IN … predicate should be allowed. These are multi-key lookup queries.
+* Examples:
+  * `SELECT … FROM table WHERE ROWKEY >= 0 AND ROWKEY < 10000`
+  * `SELECT … FROM table WHERE ROWKEY < 1000`
+  * `SELECT … FROM table WHERE ROWKEY IN (‘k0’, ‘k1’, ‘k2’)`
+
+### Tier 2 - Generic multi-shard queries
+
+While range queries may target multiple shards, expanding to generic multi-shard queries will add significant flexibility to the supported interactive query forms. Multi-shard queries should be able to target any selection of RocksDB rows via comparisons performed aginst any columns:
+
+* Arbitrary WHERE clauses should be allowed.
+* Examples:
+  * `SELECT … FROM table WHERE x = ‘x’ AND y = ‘y’`
```

Comment:
The example here, also seems to indicate we can query by any column?  not just the key column?  Are we then thinking about secondary indexes? I feel this should be Tier 4 - we could build key or ts based aggregations and joins much before generalizing the capabilities to a full fledged OLTP/RDBMS like database as Tier 2 indicates?

Reply:
Tier 0 is only key lookups. This tier proposes arbitrary `WHERE` clauses, which also implies multi-shard queries. Obviously such queries would be slow on large tables but may be worth supporting nonetheless.

If we want to do secondary indexes I would say that should go in a tier before joins, as joins probably wouldn't be feasible without secondary indexes.

Reply:
Actually, I'd still argue for doing key based joins/aggregates first, since it caters to a lot of what people can accomplish using SQL on NoSQL datastores (I know it sounds weird, but its not me :)) today using things like [CSQL](http://cassandra.apache.org/doc/latest/cql/) and [Phoenix](https://phoenix.apache.org/) . e.g: if you have a table with page_views for a user and table with user_profile, you can still build a very meaningful app by joining the two tables based on user_id 

We can the introduce secondary indexing and generalize all tiers to any secondary column (or even non-secondary if the user is willing to tolerate a full scan) as the final tier. 

what do you think?

Reply:
I think that's reasonable. If we only allow joining on row key then this is definitely feasible performance wise. The only question is, how useful is it to interactively join two relations that already have identical row keys? When this is the case the join can already be performed within a continuous query and materialized that way, so making this possible via interactive queries may not be particularly high ROI.

Reply:
yes. great point. It may still make sense, in cases where the tables are computed off different kSQL queries or user does not want to pre-materialize the computation upfront. I think the latter is the stronger case.  Same example : based on user_profile and page_views , say a website wants to dynamically render another page. 

btw I realized we focussed mostly on joins here. what do you think about aggregations? doing a simple `select user_id, count(*) as num_views from page_views group by user_id`  could have a lot more utility upfront than tackling secondary indexing ?

Reply:
Aggregates are covered in Tier 3. Or are you referring to something that isn't covered?

And I definitely agree that interactive aggregates are more useful than secondary indexing. I ultimately left secondary indexing off of this one pager because it's such a substantial undertaking. But perhaps we should put it in a tier anyways since we're basically mapping out the long-term blueprint here.

@vinothchandar what are you thinking for indexes? Something standard? e.g.,

    CREATE INDEX idx ON table (column);

Reply:
On indexing syntax (probably deserves more discussion), yes sticking to standard `CREATE INDEX . ...` would be awesome. 

> Aggregates are covered in Tier 3. Or are you referring to something that isn't covered?

yes. I was trying to ask if you think aggregates should happen before joins or generic multi-shard queries.

Reply:
I think aggregates should definitely come before joins. I'm not sure that it would make sense to do them before generic multi-shard queries though. The reason is that interactive aggregates are typically going to be grouping on columns other than the row key, otherwise there wouldn't be any need for further aggregation versus a key-based lookup. So in effect, aggregates are themselves multi-shard queries.

Are there any realistic examples of interactive aggregates that wouldn't have to scan multiple shards?

Reply:
Summarizing the discussion.. My concerns are n't really about multi shard vs single shard, but more around doing generic queries before we have any secondary indexing to make them actually usable. Any query scanning even 10s of GB scale data, would not be really interactive.  Opening up the rocksDB state to all forms of queries is a large decision and it would be good to discuss expectations upfront. 

Specifically, I was trying to understand if we can do limited forms of joins/aggregates that work with composite keys (e.g `create table pageviews (userID, pageType, ts, ... PRIMARY_KEY(userID, pageType, ts))` that stores some transformed page views data) 

-  we could first implement ability to do a `select userID, pageType, count(*) as views_per_type from pageviews where userID > "user1" AND userID < "user100" group by userID, pageType` . This pattern fits a lot of aggregations done on time-series databases
-  We could also allow joining of pageviews table with the user_profile table (mentioned above) on a range of users and then join based on any other field. i.e have a restriction on primary key being a part of the predicate for each join table.. 

Alternatively, (I think this is @derekjn 's preference) we could build Tier 2 without any secondary indexes (cuts down engineering costs a lot) with guidance that its guarantees interactive performance only with key based sql. This is true of key value stores today (and their admins actively discourage non-key based scans). 

Based on this, I am okay to keep this as-is and revisit down the line.. It would be good to capture parts of this into the doc or call out performance is out of scope for this KLIP. 

@derekjn let me know if you have more thoughts

Reply:
I agree with Derek that for most cases where people want to do joins or aggregations we would expect / encourage them to do so as persistent queries, i.e. build materialised views of the join or aggregation they need, and then perform key-based or range queries against those materialized tables.  This is going to give them the best response times.

I also agree with Vinoth that allowing filters on arbitrary rows is likely to be totally unusable in  most real world scenarios and the implications of allowing such table scans across a production system would be troubling to most SRE/DevOps teams.

However, as Derek mentions at the top, this KLIP is mainly focused on the tiered functionality that would build up the functionality of interactive queries, and is not attempting to cover the technical complexities they might entail.  So, for example, maybe we need columnar formats, or indexes, or something else to be able to build a usable tier 2,  but maybe we can hold off having those discussions for now?

Reply:
I just don't think it's worth holding off on this capability until we can ensure performance is maximized. Even then, there will still almost certainly be cases where queries are unusably slow anyways. For a lot of users naive scans will likely be fast enough, and they can expect it to improve over time as we add indexing. Also, users can work around performance issues by figuring out how to aggregate down as much as possible to minimize row count.

Another point here is that total scan time will be bounded by the largest shard, assuming we scan all shards in parallel. Scaling up the number of nodes in their cluster is therefore another workaround here, as maximum shard size will be inversely proportional to cluster size.

And as @vinothchandar has pointed out, there are also simple conditions we can detect as a first step towards optimization here. For example, if we can form a key prefix from a `WHERE` clause, we can do prefix scans instead of full table scans. If a table has the key `(x, y, z)` and we see the `WHERE` clause `WHERE x = 0 AND y = 1`, then we can do a prefix scan using the prefix `(0, 1)`. We could educate users on such internal optimizations in the docs, and even show when/if they would be applied with `EXPLAIN` support for interactive query plans:

```
ksql> EXPLAIN SELECT count(*) FROM table WHERE x = 0;
Query plan
------------
Aggregate
  -> Prefix scan (table)
ksql>
ksql> EXPLAIN SELECT count(*) FROM table WHERE y = 0;
Query plan
------------
Aggregate
-> Table scan (table)
```

For all of these reasons I believe that generic multi-shard queries are worth giving to users before doing the work to implement indexing rather than after. There are a number of interim workarounds to get usable performance for them for the larger workloads that need it, which will most likely be a minority of workloads.

@vinothchandar @big-andy-coates maybe we should do some rough math here to get a general sense of how many rows the largest shard would have to contain to make full table scans completely unusable. If we define "completely unusable" as a runtime of ~10s+ (debatable), how many rows would the largest shard need to contain to be unusable under the following conditions?

* ~100B row size
* ~1KB row size
* Modestly fast disk
* Fast SSD (most likely for serious production workloads)

My guess is that for an average row size of a few hundred bytes on fast SSDs, we'd be approaching on the order of 10 million rows in the largest shard before becoming unusable. Keeping in mind that materialized views are often going to contain aggregate data, that's a reasonably large number of rows for one shard. Even if I'm off by an order of magnitude, 1 million aggregate rows per shard is still quite a bit of headroom.

Reply:
I think this is a great discussion, with good points on all sides. I really think that we should revisit once we have more information and/or more ideas about what direction we want to take queries against materialized state.

Rather than phrasing it in terms of technical capabilities (multi-shard queries, secondary indexing), or user-visible functionality (joins or aggregates against materialized state), we should instead try to figure out the most promising use cases based on initial feedback, agree to go after it, and then develop the user experience and technical solutions for those. 

Without that as an anchor, there probably will not be any meaningful resolution to this debate.